### ISSUE TITLE
[11.0] Decimal Precision ignored on fields with `widget=monetary`

### ISSUE THREAD

**Issue by oerp-odoo** [odoo/odoo#21686]
Impacted versions:
11.0

Steps to reproduce:
Set decimal precision to be greater than 2 for some decimal accuracy and open some form that has that specific decimal accuracy used in a field with `widget="monetary"'` set. For example: set `Product Price` Decimal Accuracy to `6`. Then open product form. Both `Sales Price` and `Cost` will still show only precision of 2. Though you can see in a list view, that it has precision of 6, because in list view it does not have `widget="monetary"` used.

Current behavior:

Decimal precision is ignored when field is used with `widget="monetary"`.

Expected behavior:
Decimal precision must change with or without `monetary` widget. It is very confusing, because saved amount is correct, but when it is showed in form, it is incorrectly displayed.

**Comment by Yenthe666** [comment]
Hi @oerp-odoo,

Thanks for the report.
I've just tested this on the latest Odoo 11 and it is still the case. The form:
![image](https://user-images.githubusercontent.com/6352350/34095184-bc261c68-e3d0-11e7-83c1-ece1758287ad.png)

The treeview:
![image](https://user-images.githubusercontent.com/6352350/34095206-cff57b94-e3d0-11e7-9492-9cdf57da2cfb.png)

I'm also facing this bug report so I've reported it through the enterprise system.
OPW 800279

Regards,
Yenthe

**Comment by Ark74** [comment]
Is  https://github.com/odoo/odoo/issues/19919 related?

**Comment by Yenthe666** [comment]
Hi @oerp-odoo,

I've already got one answer on the official ticket. The answer:

> Hello,
Thank you for your inquiry. There is one additional step that you will need to complete in order for the complete price to reflect correctly, which is to set the decimal places on the currency itself. To do so, enter your Accounting app, click on Configuration from the purple bar of menu items at the top of the screen, and select Settings from the drop down menu. Scroll down to the Currencies and select Multi-Currencies in order to enable the option “Activate Other Currencies.” Search for the currency that you are using and update the decimal places per your needs (please note: you must be in developer mode). Once this is complete, your price should be reflected correctly. I am attaching a brief video of these steps for your reference.
Please let me know if this works for you or if I can provide any additional clarification. Thanks again and I hope you are having a good day!


It works a bit further then but after configuring it and editing the form view it still doesn't show the whole number. My response on the ticket:

> Hi Drew,
Thank you for the feedback and the video. I've tested this but this is still not quite what you'd expect to happen though. By having to first activate multiple currencies and then setting the rounding it feels very unnatural and overengineered. You basically have to activate features that you don't even want to use in order to get the whole sale and purchase price.
Anyways, after testing what you've said there is still an issue.
When you've configured this all and open the form view the price is shown fine (in the precision you'd like):
https://i.imgur.com/Dxcy9VT.png
Now click on edit and look at the number again, it is still shown with two numbers behind the comma:
https://i.imgur.com/6PaBFTX.png
Both in the saved form view and the edited form view it should show the full price. In edit mode it currently shows a part of the product price, even after applying these configurations.

**Comment by oerp-odoo** [comment]
@Yenthe666 Thanks for the heads up.

**Comment by Yenthe666** [comment]
Hi @oerp-odoo,

No problem. Newest feedback from Odoo:

> Hello Yenthe,
Thank you for sending over this additional information. I have looked into this and have replicated the behavior that you have mentioned on Runbot. As such, I will be escalating this ticket to our technical team for continued support. A member of our team will contact you with additional information as soon as possible.
Thanks again for your continued patience as we work to assist you with this and I hope you are having a good day. 
Best,

Looks like atleast a part of it will be fixed. I'll keep you updated.

**Comment by Yenthe666** [comment]
Fix on the way at https://github.com/odoo/odoo/pull/21900

**Comment by nim-odoo** [comment]
@ged-odoo this one is going to be open to discussion. The issue arises when a float field has the `widget='monetary'` attribute. If the float field has a decimal precision of 6 and the currency has a precision of 2, things gets a bit messed up.

At the moment, when setting `widget='monetary'`, we round to 2 decimals, i.e. the widget option has the priority over the field precision. That works for standard fields, but not for computed fields (Cost Price is computed with inverse function).

Although this behavior completely makes sense, that was not working this way in v10. One could argue that it's the behavior of v11 and screw v10, but it's not that obvious. First, most of the time the `widget='monetary'` option was set only to display the currency. That was the sole purpose of it and this is the case of the mentioned fields. Then, it is an acceptable use case to set a float field with `widget='monetary'` and expect a decimal accuracy larger than the currency accuracy. A simple example is the price of gasoline in Belgium: price per liter is always displayed with 3 decimals, while Euro is only 2 decimals. Finally, you would also have a discrepancy between a RPC call and the web client: in the reported use case, you would be able to set a 'Sales Price' of 1.123456 in RPC but not through the web client.

Several options can be considered:
- Remove `widget='monetary'`... pretty much everywhere. Because at some point, we'll get inconsistencies between the float field precision and the currency precision. We lose the currency symbol, but at least we keep the expected digit precision.
- In case of float fields with `widget='monetary'`, the digits of the field have the priority over the digits of the currency.
- Support an override of digits in the view for monetary fields. That's the solution suggested by @antonylesuisse , but that means supporting a dynamic precision, e.g.:
```
<field name="pouet" widget="monetary" digits="get_precision('Product Price')" />
```

cc @aab-odoo and maybe @rco-odoo

**Comment by Yenthe666** [comment]
@nim-odoo I saw you've just merged in https://github.com/odoo/odoo/commit/88afd055cfc96f6505eec7547419ad96e37ab28d but there is no response on your previous comment. Could you summarize what was concluded please?
@oerp-odoo https://github.com/odoo/odoo/commit/88afd055cfc96f6505eec7547419ad96e37ab28d should fix your issue probably.

**Comment by nim-odoo** [comment]
88afd055cfc96f6505eec7547419ad96e37ab28d is not related to this issue

**Comment by Yenthe666** [comment]
I've been too fast here, just another rounding issue but then for the frontend (and without the widget). My apoligies!

**Comment by antonylesuisse** [comment]
I would suggest <field name="pouet" widget="monetary" widget_options="{'digits': 'Product Price'}" /> or at least an attribute that is literal evaluable/json parsable.

**Comment by nim-odoo** [comment]
Working on it at #21920, should works with:
`<field name="pouet" widget="monetary" options="{'digits': 'Product Price'}" />`

**Comment by Yenthe666** [comment]
@nim-odoo,

Thanks for the fix!
This improves things quite a lot. Shouldn't we do this for the cost price too? See the formview:
![image](https://user-images.githubusercontent.com/6352350/34563409-ceb26122-f152-11e7-8da7-e26b547c0a93.png)
See the treeview:
![image](https://user-images.githubusercontent.com/6352350/34563425-e0ffbcbc-f152-11e7-87b6-7037fc15a969.png)

There is still a difference there.

**Comment by nim-odoo** [comment]
@Yenthe666 crap, that's because of an inherited view: https://github.com/odoo/odoo/blob/11.0/addons/stock_account/views/product_views.xml#L44
I didn't see that testing locally.

**Comment by nim-odoo** [comment]
Fixing on #22015, already working on http://290478-11-0-cf32a2.runbot13.odoo.com/web

**Comment by Yenthe666** [comment]
Yep looks fine on the runbot, so if this is merged in everything should be good!
P.S: What about the kanban view? shouldn't we apply it there too? Then the formview, the treeview and the kanban view would be consistent. Example of the kanban:
![image](https://user-images.githubusercontent.com/6352350/34565382-f321b3b6-f15a-11e7-94d6-7a23a601f8da.png)
Example of the form:
![image](https://user-images.githubusercontent.com/6352350/34565390-ff99b698-f15a-11e7-8880-858daf4f4c4a.png)

**Comment by Yenthe666** [comment]
All fine now - what do you think about the kanban view that shows another rounding @nim-odoo?

**Comment by nim-odoo** [comment]
@Yenthe666 https://github.com/odoo/odoo/pull/22178

**Comment by aymen6** [comment]
Hello I want to set total invoice 
For Example : 250.275.25 ==> To 250.275.250 
How To Do that 
![Screenshot from 2019-07-12 11-00-02](https://user-images.githubusercontent.com/24851261/61280990-c43e2300-a7b0-11e9-82fc-345c69ffbace.png)

**Comment by Yenthe666** [comment]
@aymen6 this is not the place to ask questions about development, it is for reporting issues in the official code. Ask it on the official help forums at https://www.odoo.com/forum/help-1 please :)
In short: you have a few options. Set a rounding factor, apply a rounding defined in Odoo or round it with Python code.
